src/libsync/propagateremotedeleteencryptedrootfolder.cpp

```diff
@@ -0,0 +1,191 @@
+        storeFirstErrorString(deleteJob->errorString());
+        qCWarning(PROPAGATE_REMOVE_ENCRYPTED_ROOTFOLDER) << "Delete nested item finished with error" << err << ".";
+    } else if (httpErrorCode != 204 && httpErrorCode != 404) {
+        // A 404 reply is also considered a success here: We want to make sure
+        // a file is gone from the server. It not being there in the first place
+        // is ok. This will happen for files that are in the DB but not on
+        // the server or the local file system.
+
+        // Normally we expect "204 No Content"
+        // If it is not the case, it might be because of a proxy or gateway intercepting the request, so we must
+        // throw an error.
+        storeFirstErrorString(tr("Wrong HTTP code returned by server. Expected 204, but received \"%1 %2\".")
+                       .arg(httpErrorCode)
+                       .arg(deleteJob->reply()->attribute(QNetworkRequest::HttpReasonPhraseAttribute).toString()));
+        if (_item->_httpErrorCode == 0) {
+            _item->_httpErrorCode = httpErrorCode;
+        }
+
+        qCWarning(PROPAGATE_REMOVE_ENCRYPTED_ROOTFOLDER) << "Delete nested item finished with error" << httpErrorCode << ".";
+    }
+
+    if (_nestedItems.size() == 0) {
+        // we wait for all _nestedItems' DeleteJobs to finish, and then - fail if any of those jobs has failed
+        if (networkError() != QNetworkReply::NetworkError::NoError || _item->_httpErrorCode != 0) {
+            const int errorCode = networkError() != QNetworkReply::NetworkError::NoError ? networkError() : _item->_httpErrorCode;
+            qCCritical(PROPAGATE_REMOVE_ENCRYPTED_ROOTFOLDER) << "Delete of nested items finished with error" << errorCode << ". Failing the entire sequence.";
+            taskFailed();
+            return;
+        }
+        unlockFolder();
+    }
+}
+
+void PropagateRemoteDeleteEncryptedRootFolder::deleteNestedRemoteItem(const QString &filename)
+{
+    qCInfo(PROPAGATE_REMOVE_ENCRYPTED_ROOTFOLDER) << "Deleting nested encrypted remote item" << filename;
+
+    auto deleteJob = new DeleteJob(_propagator->account(), _propagator->fullRemotePath(filename), this);
+    deleteJob->setFolderToken(_folderToken);
+    deleteJob->setProperty(encryptedFileNamePropertyKey, filename);
+
+    connect(deleteJob, &DeleteJob::finishedSignal, this, &PropagateRemoteDeleteEncryptedRootFolder::slotDeleteNestedRemoteItemFinished);
+
+    deleteJob->start();
+}
+
+void PropagateRemoteDeleteEncryptedRootFolder::decryptAndRemoteDelete()
+{
+    auto job = new OCC::SetEncryptionFlagApiJob(_propagator->account(), _item->_fileId, OCC::SetEncryptionFlagApiJob::Clear, this);
+    connect(job, &OCC::SetEncryptionFlagApiJob::success, this, [this] (const QByteArray fileId) {
+        Q_UNUSED(fileId);
+        deleteRemoteItem(_item->_file);
+    });
+    connect(job, &OCC::SetEncryptionFlagApiJob::error, this, [this] (const QByteArray fileId, int httpReturnCode) {
```

Comment:
Missing ref on fileId.

Reply:
@er-vin In fact the ref is missing in the **SetEncryptionFlagApiJob** declaration

```
void success(const QByteArray fileId);
void error(const QByteArray fileId, int httpReturnCode);
```

Reply:
@er-vin should I fix that in the **clientsideencryptionjobs.h** ?

Reply:
*sigh* yes, please fix it there as well.

Reply:
> Couple of nitpicks, then clean up the history as you see fit and we're good to go I'd say.

@er-vin done